Implemented Adaptsize #103

muhhae opened the pull request:
# Overview
Implemented adaptsize admission algorithm. #103 
admission param:
- reconf-interval, default = 30000, suitable for cloudPhysicsIO.vcsi
- max-iteration, default = 15
# Result
## Using cloudPhysicsIO.vscsi 
### AdaptSize
![Adaptsize](https://github.com/user-attachments/assets/e115088e-5e50-424d-9e6f-b407248ec5b3)
### Default
![Default](https://github.com/user-attachments/assets/994ad68d-483d-4a69-ba95-21d83e5be1f6)
### BloomFilter
![BloomFilter](https://github.com/user-attachments/assets/81543329-3db3-40f3-8eca-560559d9eae5)
### SizeProb
![SizeProb](https://github.com/user-attachments/assets/dfad3a6d-6160-40dc-b7b2-8add37896187)
Result shows that AdaptSize better at any cache size compared to bloomfilter. But perform worse at larger size compared to default admission and maintained high miss-ratio. 
My hypothesis would be that the data is just too small for adaptsize to take effect. So I tried to use larger dataset I choose cluster50 from https://ftp.pdl.cmu.edu/pub/datasets/twemcacheWorkload/cacheDatasets/twitter/.
## Using cluster50
### AdaptSize
![big_adaptsize](https://github.com/user-attachments/assets/cc4482d0-3774-47e0-b71e-d1c5fc96b5ce)
#### Default
![big_default](https://github.com/user-attachments/assets/21749797-737c-4f83-8a39-173a5c7a5d86)
### BloomFilter
![big_bloomfilter](https://github.com/user-attachments/assets/c9eddf73-5fc7-459e-835d-e22632e4a4f6)
### SizeProb
![big_sizeprob](https://github.com/user-attachments/assets/98f06b12-f57c-445b-ad22-bdb883cca95b)
Result show that on larger dataset adaptsize achieves really big improvement compared to other on smaller cache size while maintaining slightly bigger improvement on larger cache size.
# Known Issues
- ~~Choosing reconf-interval too small would result in segmentation fault(possible data structure/memory leak issue, I couldn't find it) while too big reconf-interval would make it useless. Sweet spot is 3/10 of data size (original implementation use 250k).~~ [Fixed] 
# Possible Improvement
- Use in-house data structure instead of std::unordered_map and std::vector
# Notes:
- Is there like look up callback for admissioner, currently I put admissioner->update_stats on cache_get_base(). 
- getting cache size as admissioner, currently i fetch it on update to get current cache_size.

muhhae committed:
Adaptsize implementation

muhhae committed:
AdaptSize param parsing and initialization

muhhae committed:
Implementing AdaptSize with C++ instead of C for better data structure

muhhae committed:
fix cmake

muhhae committed:
docs

muhhae committed:
fix formula

muhhae committed:
Merge branch '1a1a11a:develop' into develop

1a1a11a commented:
Awesome work! I think using c++ is ok, no need to change to internal structures. 
a few suggestions
1. can you add test for adaptsize and other algorithms? 
2. can you find out why there is segfault. A tip: using `CMAKE_BUILD_TYPE-Debug` in the build can provide more information when you debug it, and Valgrind is a useful tool.

muhhae commented:
> Awesome work! I think using c++ is ok, no need to change to internal structures. a few suggestions
> 
>     1. can you add test for adaptsize and other algorithms?
> 
>     2. can you find out why there is segfault. A tip: using `CMAKE_BUILD_TYPE-Debug` in the build can provide more information when you debug it, and Valgrind is a useful tool.

1. Sure, may I know what you would expect from the test?
2. I've tried using Valgrind without CMAKE_BUILD_TYPE-Debug, and its just stuck, I'm gonna try it again thanks.

muhhae committed:
refactor and restructure code for easier understanding

muhhae committed:
fix segfault problem

muhhae committed:
docs and use consistent indent

muhhae commented:
I've fixed segfault issue, 
below is the result with different reconf_interval. Smaller reconf_interval tend to perform better at smaller cache. 
#### 30M reconf_interval
![30M_reconf_interval](https://github.com/user-attachments/assets/339d172b-81c8-43f4-ba7b-ad7066d9d5a2)
#### 3M reconf_interval
![3M_reconf_interval](https://github.com/user-attachments/assets/85cdb173-43d7-4055-b245-50cc3cee2e07)
#### 300K reconf_interval
![300K_reconf_interval](https://github.com/user-attachments/assets/b9272581-ea6f-4a23-b648-63bb401c5b48)

1a1a11a commented:
1. the test is to make sure that if someone makes changes to this and other code, we want to make sure it will not break the implementation. 
2. `valgrind` is very slow and can take hours, but if the segfault happens early, it should not take too much time.

1a1a11a commented:
One more request - The current output all show LRU, can you find a way to modify the name to include the admission algorithm and parameters in the name?

muhhae commented:
> One more request - The current output all show LRU, can you find a way to modify the name to include the admission algorithm and parameters in the name?

Sure, including eviction param or just admission param?

1a1a11a commented:
Let's include both. You can think about the following scenarios where you run 100000s of simulations in a cluster, the output are all logged, how would you distinguish between different parameters?

1a1a11a commented:
We also need to have tests to make sure other changes do not affect the implementation.

muhhae committed:
include admission name and param to the output

muhhae commented:
> Let's include both. You can think about the following scenarios where you run 100000s of simulations in a cluster, the output are all logged, how would you distinguish between different parameters?

how about this?
![image](https://github.com/user-attachments/assets/fb74bbfc-11cd-4b65-b2b1-a53fade20518)

![image](https://github.com/user-attachments/assets/fde18930-0e0b-4879-925c-30e4f4c08e1e)

![image](https://github.com/user-attachments/assets/50420c34-1dbe-4514-a681-9f7e044265f6)

I used param from user so if they don't give any param it just look like lru[], I think if we're gonna include default param, we need to change every algo implementation that already exists. So I haven't done that (Its possible).

muhhae committed:
Revert "include admission name and param to the output"

This reverts commit 03fea8c24d65be0cc4753112c368e8d01e7bb59d.

muhhae committed:
added test for AdaptSize,Size,SizeProb

muhhae commented:
> > Let's include both. You can think about the following scenarios where you run 100000s of simulations in a cluster, the output are all logged, how would you distinguish between different parameters?
> 
> how about this? ![image](https://private-user-images.githubusercontent.com/115412148/418361394-fb74bbfc-11cd-4b65-b2b1-a53fade20518.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NzQ5MzgsIm5iZiI6MTc0MDk3NDYzOCwicGF0aCI6Ii8xMTU0MTIxNDgvNDE4MzYxMzk0LWZiNzRiYmZjLTExY2QtNGI2NS1iMmIxLWE1M2ZhZGUyMDUxOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwM1QwNDAzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNjJiNjM0ZTY5MWMxZGY0ZGQ2Yjc4MGI5ZmM1OGQ1ZGQ5NTU3MmY1ZWRmNTY3MTVmNjY5NmUxZWQwZmE1NDMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iMvLKDUT3oyr0tfAOnXweQQnU6SJ7fd4_ahf6YtrEjc)
> 
> ![image](https://private-user-images.githubusercontent.com/115412148/418361405-fde18930-0e0b-4879-925c-30e4f4c08e1e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NzQ5MzgsIm5iZiI6MTc0MDk3NDYzOCwicGF0aCI6Ii8xMTU0MTIxNDgvNDE4MzYxNDA1LWZkZTE4OTMwLTBlMGItNDg3OS05MjVjLTMwZTRmNGMwOGUxZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwM1QwNDAzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTViYmYyNDBmYjc5ZDU5MTlhZTdiYTNmYTkwODhhMTZmN2E5MjUxMDdlM2M0YzMwMzdhM2Q1YTA2NGFkNzdhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.dlxOPjfGMff0Y53lKXbsEKre5-xNdXHD5rWVvRsJELw)
> 
> ![image](https://private-user-images.githubusercontent.com/115412148/418361689-50420c34-1dbe-4514-a681-9f7e044265f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NzQ5MzgsIm5iZiI6MTc0MDk3NDYzOCwicGF0aCI6Ii8xMTU0MTIxNDgvNDE4MzYxNjg5LTUwNDIwYzM0LTFkYmUtNDUxNC1hNjgxLTlmN2UwNDQyNjVmNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwM1QwNDAzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZjExZjFlMTAxYTg2OWFjNDdiYmIwOTgyMGEzYjVjMWRmM2UxNWVlNDdkYTZiYTVjNTMwNWY4OTg3MDA2NThmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FjqeTnoLxGBzqVzKDT8emiCD9vWT5ZY-0U9vharQx8M)
> 
> I used param from user so if they don't give any param it just look like lru[], I think if we're gonna include default param, we need to change every algo implementation that already exists. So I haven't done that (Its possible).

~~I reverted this commit, because it result in segfault when we do test (Currently investigating it)~~[fixed].
 
I've successfully created test for AdaptSize, Size, SizeProb.

~~I disabled BloomFilter Test because it caused SegFault[when ran together with other algo, running it alone turns out fine], 
segfault happened on this line~~[Fixed]
```C
// test_admissionAlgo.c:98
  cache_stat_t *res = simulate_at_multi_sizes_with_step_size(reader, cache, STEP_SIZE, NULL, 0, 0, _n_cores(), false);
```
~~The other is fine.~~[Fixed]

muhhae committed:
bloomfilter test data

muhhae committed:
Reapply "include admission name and param to the output"

This reverts commit 850f5f0dfa1468f90a8b2c9542f39a8e762880d1.

muhhae committed:
fix segfault issues when applying cache param and name to output

muhhae committed:
enabled bloomFilter test

muhhae commented:
All Done

muhhae committed:
Merge branch 'develop' into develop

muhhae committed:
use smaller array size

assuming cache_param = 256, admission_param = 256, cache_name = 64,
admission_name = 64, total = 640

1a1a11a committed:
Merge branch 'develop' into develop

1a1a11a commented:
I just realized that the admission algorithm was created as an eviction algorithm, we should use `-a` for admission so that we can use different eviction algorithms with admission. Can you update this?

1a1a11a commented:
@muhhae You can create a new PR for this.

muhhae commented:
> I just realized that the admission algorithm was created as an eviction algorithm, we should use `-a` for admission so that we can use different eviction algorithms with admission. Can you update this? 

I don't really understand what you mean, as I already use -a for adaptsize could you elaborate more.

1a1a11a commented:
Oh, the changes are in `test_common.h`, and I thought that it is cache_init.h 
Please ignore the comment. You did it correctly
